salesforce/bulkwrite.go

```diff
@@ -0,0 +1,299 @@
+package salesforce
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/amp-labs/connectors/common"
+	"github.com/spyzhov/ajson"
+)
+
+// BulkWriteParams defines how we are writing data to a SaaS API.
+type BulkWriteParams struct {
+	// The name of the object we are writing, e.g. "Account"
+	ObjectName string // required
+
+	// The external ID of the object instance we are updating. Provided in the case of UPDATE, but not CREATE.
+	ExternalId string // required
+
+	// The path to the CSV file we are writing
+	FilePath string // required
+}
+
+var (
+	ErrKeyNotFound     = errors.New("key not found")
+	ErrUnknownNodeType = errors.New("unknown node type when parsing JSON")
+	ErrInvalidType     = errors.New("invalid type")
+	ErrInvalidJobState = errors.New("invalid job state")
+)
+
+// BulkWriteResult is what's returned from writing data via the BulkWrite call.
+type BulkWriteResult struct {
+	// State is the state of the bulk job process
+	State string `json:"state"`
+	// JobId is the ID of the bulk job process
+	JobId string `json:"jobId"`
+}
+
+func (c *Connector) BulkWrite( //nolint:funlen,cyclop
+	ctx context.Context,
+	config BulkWriteParams,
+) (*BulkWriteResult, error) {
+	// cretes batch upload job, returns json with id and other info
+	res, err := c.createJob(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("createJob failed: %w", err)
+	}
+
+	jobCreateRes, err := ParseAjsonNodeToMap(res)
+	if err != nil {
+		return nil, fmt.Errorf("parsing result of createJob failed: %w", errors.Join(err, common.ErrParseError))
+	}
+
+	state, ok := jobCreateRes["state"].(string) //nolint:varnamelen
```

Comment:
ajson already offers pretty good helper functions.

For example there's the [GetObject](https://pkg.go.dev/github.com/spyzhov/ajson#Node.GetObject) function which will turn res into a map of nodes (which replaces the need for the ParseAjsonNodeToMap function), and [GetString](https://pkg.go.dev/github.com/spyzhov/ajson#Node.GetString) function for getting the string value for each key.
```
nodesMap, err := res.GetObject("state")
if err != nil { ... }
state, err := s["state"].GetString()
if err != nil { ...}
```

This is safe to do and will not panic even if the keys don't exist, I tried this out ([see playground](https://play.golang.com/p/DZQAYprdsix))